rotkehlchen/chain/evm/decoding/cowswap/decoder.py

```diff
@@ -328,20 +317,35 @@ def _aggregator_post_decoding(
 
         return decoded_events
 
+    def _match_cowswap_counterparty(self, context: DecoderContext) -> DecodingOutput:
+        """Sets the CowSwap counterparty for events involving the settlement contract.
+
+        This is needed because smart accounts may route through CowSwap without
+        the transaction's to_address being the settlement contract directly.
+        """
+        for event in context.decoded_events:
+            if event.address == self.settlement_address:
+                return DecodingOutput(matched_counterparty=CPT_COWSWAP)
+
+        return DEFAULT_DECODING_OUTPUT
+
     # -- DecoderInterface methods
 
     @staticmethod
     def counterparties() -> tuple[CounterpartyDetails, ...]:
         return (COWSWAP_CPT_DETAILS,)
 
     def addresses_to_decoders(self) -> dict[ChecksumEvmAddress, tuple[Any, ...]]:
-        return {self.native_asset_flow_address: (self._decode_native_asset_orders,)}
+        return {
+            self.native_asset_flow_address: (self._decode_native_asset_orders,),
+            self.settlement_address:  (self._match_cowswap_counterparty,),
+        }
```

Comment:
something odd here. `addresses_to_decoders` has `settlement_address` so it means that there is an event from where `settlement_address` emits it (basically the condition that you are iterating in the loop.

That means either that you can avoid the post decoding by calling the post decoding logic directly instead or you just need to do a return if that is not possible

Reply:
Yabir says he resolved this due to:

 the post decoding is needed because it has only one leg of the trade and the other comes later. It is stated in the comment of the function.
We can't use the mapping of address to counterparty because the decoding logic matches against the to_address of the transaction and not the event.address